nnet_utils/nnet_conv.h

```diff
@@ -31,86 +32,136 @@ struct conv_config
 
 template<class data_T, class res_T, typename CONFIG_T>
 void conv_1d(
 	     data_T    data[CONFIG_T::y_in][CONFIG_T::n_chan],
-	     res_T     res[CONFIG_T::y_in][CONFIG_T::n_chan],
-	     typename CONFIG_T::weight_t  weights[CONFIG_T::y_filt][CONFIG_T::n_chan],
-	     typename CONFIG_T::bias_t    biases[CONFIG_T::n_chan])
+	     res_T     res[CONFIG_T::y_out][CONFIG_T::n_filt],
+	     typename CONFIG_T::weight_t  weights[CONFIG_T::y_filt][CONFIG_T::n_chan][CONFIG_T::n_filt],
+	     typename CONFIG_T::bias_t    biases[CONFIG_T::n_filt])
 {
-    // conv_1d: 1-dimensional convolution
-    //   - Also includes multiple input channels
-    //   - Only allows new data on each ROW (i.e. this is NOT a 2D convolution)
-    // Only ONE output channel per input channel
-
-    // Initial directives used from HLS User guide, pg 381
-    // (https://www.xilinx.com/support/documentation/sw_manuals/xilinx2015_4/ug902-vivado-high-level-synthesis.pdf)
-
-    // TODO: Figure out how to correctly pipeline FiltLoop-- It sort of needs to be pipelined
-    // across iterations of ChanLoop. Otherwise it does not want to consistently hit 10 ns timing
-    data_T buffer[CONFIG_T::y_filt][CONFIG_T::n_chan];
-    for(int ii = 0; ii < CONFIG_T::y_filt; ii++) {
-      for(int chan = 0; chan < CONFIG_T::n_chan; chan++){
-      #pragma HLS UNROLL
-	// Initialize buffer to zero: effecively a form of "same" zero padding
-	buffer[ii][chan] = 0;
-      }
+
+    typename CONFIG_T::accum_t mult[CONFIG_T::y_out][CONFIG_T::n_filt][CONFIG_T::n_chan][CONFIG_T::y_filt];
+    typename CONFIG_T::accum_t acc[CONFIG_T::y_out][CONFIG_T::n_filt];
+
+    #pragma HLS ARRAY_PARTITION variable=mult complete
+    #pragma HLS ARRAY_PARTITION variable=acc complete
+    
+    // Use a function_instantiate in case it helps to explicitly optimize unchanging weights/biases 
+    #pragma HLS function_instantiate variable=weights,biases
+    
+    // Parallel mode
+    #pragma HLS PIPELINE
+    #pragma HLS ARRAY_PARTITION variable=biases complete
+  
+    // Limit multipliers to control parallelization
+    int multiplier_limit = ceil( (CONFIG_T::y_out*CONFIG_T::n_filt*CONFIG_T::n_chan*CONFIG_T::y_filt - 
+	                          CONFIG_T::n_filt*CONFIG_T::n_chan*CONFIG_T::pad_left - 
+                              	  CONFIG_T::n_filt*CONFIG_T::n_chan*CONFIG_T::pad_right) 
+	                        / CONFIG_T::reuse_factor ); 
+    #pragma HLS ALLOCATION instances=mul limit=multiplier_limit operation
+
+    
+    // Convolve, saving all multiplication results to accumulate later
+    ConvOut: for(int ii = 0; ii < CONFIG_T::y_out; ii++) {
+        ConvFilt: for(int ff = 0; ff < CONFIG_T::n_filt; ff++){
+            ConvChan: for(int cc = 0; cc < CONFIG_T::n_chan; cc++){
+		
+		//Multiply
+                ConvMult: for(int jj = 0; jj < CONFIG_T::y_filt; jj++){
+                    if((ii*CONFIG_T::stride+jj) < CONFIG_T::pad_left || (ii*CONFIG_T::stride+jj) >= (CONFIG_T::pad_left + CONFIG_T::y_in)){
```

Comment:
should the second pad_left be a pad_right?

Reply:
We think this is correct.  @jmduarte explained in your office, but just to update the conversation here -- you can imagine the loop being over the size of the padded data, and this tells you to skip the left pad (i*stride+j < pad_left) and skip everything after the non-padded data (i*stride+j >= pad_left+data) ie the right pad.  It could also be written in terms of pad_right, but this way works too.